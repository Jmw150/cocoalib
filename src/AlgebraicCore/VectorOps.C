//   Copyright (c)  2016  John Abbott and Anna M. Bigatti

//   This file is part of the source of CoCoALib, the CoCoA Library.
//
//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.
//
//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.

#include "CoCoA/VectorOps.H"

#include "CoCoA/BigIntOps.H"
#include "CoCoA/utils.H"

#include<list>
using std::list;
#include<vector>
using std::vector;

namespace CoCoA
{

  // This should be a template fn?
  BigInt BinaryProduct(vector<BigInt>& v)
  {
    while (true)
    {
      const int n = len(v);
      if (n < 4) break;
      if (n&1) v[n-2] *= v[n-1];
      const int n2 = n/2; // integer division!
      for (int i=0; i < n2; ++i)
        v[i] = v[2*i]*v[2*i+1];
      v.resize(n2);
    }
    const int n = len(v);
    if (n == 1) return v[0];
    if (n == 2) return v[1]*v[1];
    return v[0]*v[1]*v[2];
  }

  BigInt BinaryProduct(const list<BigInt>& L)
  {
    if (L.empty()) CoCoA_THROW_ERROR(ERR::Empty, "BinaryProduct");
    if (len(L) == 1) return L.front();
    if (len(L) == 2) return L.front()*L.back();
    list<BigInt> tmp;
    const auto END = L.end();
    for (auto it = L.begin(); it != END; ++it)
    {
      const auto prev = it;
      ++it;
      if (it == END) { tmp.back() *= *it; break; }
      tmp.push_back((*prev)*(*it));
    }
    return BinaryProduct(tmp);
  }
  
} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/VectorOps.C,v 1.7 2022/02/18 14:12:01 abbott Exp $
// $Log: VectorOps.C,v $
// Revision 1.7  2022/02/18 14:12:01  abbott
// Summary: Updated copyright notice (now restrictive; see redmine 1555)
//
// Revision 1.6  2021/01/07 15:16:53  abbott
// Summary: Corrected copyright
//
// Revision 1.5  2020/06/17 15:49:29  abbott
// Summary: Changed CoCoA_ERROR into CoCoA_THROW_ERROR
//
// Revision 1.4  2020/02/18 11:30:16  abbott
// Summary: Some tidying; change var name l in to L
//
// Revision 1.3  2018/05/22 14:16:40  abbott
// Summary: Split BigRat into BigRat (class defn + ctors) and BigRatOps
//
// Revision 1.2  2018/05/18 12:24:47  bigatti
// -- renamed IntOperations --> BigIntOps
//
// Revision 1.1  2018/05/17 15:24:30  bigatti
// -- renamed VectorOperations --> VectorOps
//
// Revision 1.1  2017/02/01 12:17:57  abbott
// Summary: Added new fn BinaryProduct -- never called currently
//
//
