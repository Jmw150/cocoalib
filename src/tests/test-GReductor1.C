//   Copyright (c)  2007,2017  John Abbott and Anna M. Bigatti

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BigIntOps.H"
#include "CoCoA/BuildInfo.H"
#include "CoCoA/FractionField.H"
#include "CoCoA/FreeModule.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/ModuleOrdering.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/RingDistrMPolyClean.H"
#include "CoCoA/RingDistrMPolyInlFpPP.H"
#include "CoCoA/RingDistrMPolyInlPP.H"
#include "CoCoA/RingFp.H"
#include "CoCoA/RingQQ.H"
#include "CoCoA/SparsePolyRing.H"
#include "CoCoA/TmpGOperations.H"
#include "CoCoA/TmpGPoly.H"
#include "CoCoA/VectorOps.H"
#include "CoCoA/submodule.H"
#include "CoCoA/symbol.H"

#include <algorithm>
using std::find_if;
#include <functional>
using std::not1;
using std::ptr_fun;
#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
using std::boolalpha;
#include <string>
#include <vector>
using std::vector;

//----------------------------------------------------------------------
// Test for Buchberger's Algorithm 
// functions: ComputeGBasis
// environments: DMP (Fp, Q), DMPI (Fp), DMPII
//----------------------------------------------------------------------
namespace CoCoA
{

  void PrintInfo(const std::string& s, const SparsePolyRing& P, const vector<RingElem>& InputPolys)
  {
    cout << "Executing test " << s << endl << P << "   ";
    cout << "GradingDim = " << GradingDim(P) << "   ";
    if (GradingDim(P)!=0)
      cout << "IsHomog(InputPolys) = " << IsHomog(InputPolys);
    cout << endl;
  }


  // expecting P to have 5 indeterminates
  void testC4_h(SparsePolyRing P)
  {
    RingElem x = indet(P,0),  y = indet(P,1),  z = indet(P,2), t=indet(P,3);
    RingElem h = indet(P,4);
    PolyList InputPolys,GB;
    InputPolys.push_back(x+y+z+t);
    InputPolys.push_back(x*y+y*z+z*t+t*x);
    InputPolys.push_back(x*y*z+y*z*t+z*t*x+t*x*y);
    InputPolys.push_back(x*y*z*t-power(h,4));
    PrintInfo("C4_h", P, InputPolys);
    PolyList MinGens;
    ComputeGBasis(GB, MinGens, InputPolys);
    CoCoA_ASSERT_ALWAYS(len(MinGens)==4);
    monic(GB);
    cout << "C4_h = " << GB << "\n"<<endl;
  }


  void testC4(SparsePolyRing P)
  {
    RingElem x = indet(P,0),  y = indet(P,1),  z = indet(P,2), t = indet(P,3);
    PolyList InputPolys,GB;
    InputPolys.push_back(x+y+z+t);
    InputPolys.push_back(x*y + y*z + z*t + t*x);
    InputPolys.push_back(x*y*z + y*z*t + z*t*x + t*x*y);
    InputPolys.push_back(x*y*z*t - 1);
    PrintInfo("C4", P, InputPolys);
    PolyList MinGens;
    ComputeGBasis(GB, MinGens, InputPolys);
    CoCoA_ASSERT_ALWAYS(len(MinGens)==0);
    monic(GB);
    cout << "C4 = " << GB << "\n"<<endl;
  }

  // expecting P to have 4 indeterminates
  void testCurve_h(SparsePolyRing P)
  {
    RingElem t = indet(P,0),  x = indet(P,1),  y = indet(P,2), z=indet(P,3);
    PolyList InputPolys,GB;
    InputPolys.push_back(power(t,3)-power(x,3));
    InputPolys.push_back(power(t,5)-power(y,5));
    InputPolys.push_back(power(t,7)-power(z,7));
    PrintInfo("Curve_h", P, InputPolys);
    PolyList MinGens;
    ComputeGBasis(GB, MinGens, InputPolys);
    CoCoA_ASSERT_ALWAYS(len(MinGens)==3);
    monic(GB);
    cout << "Curve_h = " << GB << "\n"<<endl;
  }

  // expecting P to have 5 indeterminates
  void testCurve(SparsePolyRing P)
  {
    const vector<RingElem>& x = indets(P);
    PolyList InputPolys, GB;
    InputPolys.push_back(power(x[0],3) - x[1]);
    InputPolys.push_back(power(x[0],5) - x[2]);
    InputPolys.push_back(power(x[0],7) - x[3]);
    PrintInfo("Curve", P, InputPolys);
    PolyList MinGens;
    ComputeGBasis(GB, MinGens, InputPolys);
    CoCoA_ASSERT_ALWAYS(len(MinGens)==0);
    monic(GB);
    cout << "Curve = " << GB << "\n" << endl;
  }

  void program()
  {
    GlobalManager CoCoAFoundations(UseNonNegResidues);
    cout << boolalpha;
  
    // This is a test of the correctness of GB computations for the three cases:
    // AllGraded, AllAffine and GradedRing, Inhomogenous Input for ideals only.
    // This does not check if the computations have been done in the right way,
    // only if the result is correct.

    // the results have been tested with cocoa 4
  
    ring ZZmod = NewZZmod(101);  
  
    // Cyclic 4
    const vector<symbol> X = SymbolRange("x", 0, 4);
    SparsePolyRing P(NewPolyRing_DMPI(ZZmod, X));
    SparsePolyRing Q(NewPolyRing_DMPII(ZZmod, X));
    SparsePolyRing R(NewPolyRing(RingQQ(), NewSymbols(5)));
    SparsePolyRing T(NewPolyRing_DMPI(RingQQ(), X));

    // Graded Ring, Homogeneous Input
    testC4_h(P);
    testC4_h(Q);
    testC4_h(R);
    testC4_h(T);
 
    // Graded Ring, Non-homogeneous Input
    testC4(P);
    testC4(Q);
    testC4(R);
    testC4(T);
  
    SparsePolyRing LexR = NewPolyRing(RingQQ(), X, lex);
    // Non-graded Ring, Non-homogeneous Input
    //testC4(LexR,"NewPolyRing Q");
    
    // Curve
    const vector<symbol> Y = SymbolRange("x", 0 ,3);
    SparsePolyRing P1(NewPolyRing_DMPI(ZZmod, Y));
    SparsePolyRing Q1(NewPolyRing_DMPII(ZZmod, Y));
    SparsePolyRing R1(NewPolyRing(RingQQ(), NewSymbols(4)));
    SparsePolyRing T1 (NewPolyRing_DMPI(RingQQ(), Y));

    // Graded Ring, Homogeneous Input
    testCurve_h(P1);
    testCurve_h(Q1);
    testCurve_h(R1);
    testCurve_h(T1);
  
    // Graded Ring, Non-homogeneous Input
    testCurve(P1);
    testCurve(Q1);
    testCurve(R1);
    testCurve(T1);
  
    SparsePolyRing LexR1 = NewPolyRing(RingQQ(), Y, lex);

    // Non-graded Ring, Non-homogeneous Input
    testCurve(LexR1);
  }

} // end of namespace CoCoA


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    CoCoA::program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  CoCoA::BuildInfo::PrintAll(cerr);
  return 1;
}
