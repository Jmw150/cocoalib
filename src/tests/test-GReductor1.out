Executing test C4_h
RingWithID(3, "ZZ/(101)[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = true
C4_h = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] +100*x[1]*x[3]^2 +100*x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 +100*x[1]*x[3]^3 +x[2]*x[3]^3 +100*x[3]^4 +100*x[4]^4,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 +100*x[2]*x[4]^4 +100*x[3]*x[4]^4,  x[1]*x[3]^4 +x[3]^5 +100*x[1]*x[4]^4 +100*x[3]*x[4]^4,  x[2]^2*x[3]^4 +x[1]*x[2]*x[4]^4 +100*x[1]*x[3]*x[4]^4 +x[2]*x[3]*x[4]^4 +99*x[3]^2*x[4]^4]

Executing test C4_h
RingWithID(4, "ZZ/(101)[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = true
C4_h = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] +100*x[1]*x[3]^2 +100*x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 +100*x[1]*x[3]^3 +x[2]*x[3]^3 +100*x[3]^4 +100*x[4]^4,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 +100*x[2]*x[4]^4 +100*x[3]*x[4]^4,  x[1]*x[3]^4 +x[3]^5 +100*x[1]*x[4]^4 +100*x[3]*x[4]^4,  x[2]^2*x[3]^4 +x[1]*x[2]*x[4]^4 +100*x[1]*x[3]*x[4]^4 +x[2]*x[3]*x[4]^4 +99*x[3]^2*x[4]^4]

Executing test C4_h
RingWithID(5, "QQ[#[0],#[1],#[2],#[3],#[4]]")   GradingDim = 1   IsHomog(InputPolys) = true
C4_h = [#[0] +#[1] +#[2] +#[3],  #[1]^2 +2*#[1]*#[3] +#[3]^2,  #[1]*#[2]^2 +#[2]^2*#[3] -#[1]*#[3]^2 -#[3]^3,  #[1]*#[2]*#[3]^2 +#[2]^2*#[3]^2 -#[1]*#[3]^3 +#[2]*#[3]^3 -#[3]^4 -#[4]^4,  #[2]^3*#[3]^2 +#[2]^2*#[3]^3 -#[2]*#[4]^4 -#[3]*#[4]^4,  #[1]*#[3]^4 +#[3]^5 -#[1]*#[4]^4 -#[3]*#[4]^4,  #[2]^2*#[3]^4 +#[1]*#[2]*#[4]^4 -#[1]*#[3]*#[4]^4 +#[2]*#[3]*#[4]^4 -2*#[3]^2*#[4]^4]

Executing test C4_h
RingWithID(6, "QQ[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = true
C4_h = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] -x[1]*x[3]^2 -x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 -x[1]*x[3]^3 +x[2]*x[3]^3 -x[3]^4 -x[4]^4,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 -x[2]*x[4]^4 -x[3]*x[4]^4,  x[1]*x[3]^4 +x[3]^5 -x[1]*x[4]^4 -x[3]*x[4]^4,  x[2]^2*x[3]^4 +x[1]*x[2]*x[4]^4 -x[1]*x[3]*x[4]^4 +x[2]*x[3]*x[4]^4 -2*x[3]^2*x[4]^4]

Executing test C4
RingWithID(3, "ZZ/(101)[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = false
C4 = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] +100*x[1]*x[3]^2 +100*x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 +100*x[1]*x[3]^3 +x[2]*x[3]^3 +100*x[3]^4 +100,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 +100*x[2] +100*x[3],  x[1]*x[3]^4 +x[3]^5 +100*x[1] +100*x[3],  x[2]^2*x[3]^4 +x[1]*x[2] +100*x[1]*x[3] +x[2]*x[3] +99*x[3]^2]

Executing test C4
RingWithID(4, "ZZ/(101)[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = false
C4 = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] +100*x[1]*x[3]^2 +100*x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 +100*x[1]*x[3]^3 +x[2]*x[3]^3 +100*x[3]^4 +100,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 +100*x[2] +100*x[3],  x[1]*x[3]^4 +x[3]^5 +100*x[1] +100*x[3],  x[2]^2*x[3]^4 +x[1]*x[2] +100*x[1]*x[3] +x[2]*x[3] +99*x[3]^2]

Executing test C4
RingWithID(5, "QQ[#[0],#[1],#[2],#[3],#[4]]")   GradingDim = 1   IsHomog(InputPolys) = false
C4 = [#[0] +#[1] +#[2] +#[3],  #[1]^2 +2*#[1]*#[3] +#[3]^2,  #[1]*#[2]^2 +#[2]^2*#[3] -#[1]*#[3]^2 -#[3]^3,  #[1]*#[2]*#[3]^2 +#[2]^2*#[3]^2 -#[1]*#[3]^3 +#[2]*#[3]^3 -#[3]^4 -1,  #[2]^3*#[3]^2 +#[2]^2*#[3]^3 -#[2] -#[3],  #[1]*#[3]^4 +#[3]^5 -#[1] -#[3],  #[2]^2*#[3]^4 +#[1]*#[2] -#[1]*#[3] +#[2]*#[3] -2*#[3]^2]

Executing test C4
RingWithID(6, "QQ[x[0],x[1],x[2],x[3],x[4]]")   GradingDim = 1   IsHomog(InputPolys) = false
C4 = [x[0] +x[1] +x[2] +x[3],  x[1]^2 +2*x[1]*x[3] +x[3]^2,  x[1]*x[2]^2 +x[2]^2*x[3] -x[1]*x[3]^2 -x[3]^3,  x[1]*x[2]*x[3]^2 +x[2]^2*x[3]^2 -x[1]*x[3]^3 +x[2]*x[3]^3 -x[3]^4 -1,  x[2]^3*x[3]^2 +x[2]^2*x[3]^3 -x[2] -x[3],  x[1]*x[3]^4 +x[3]^5 -x[1] -x[3],  x[2]^2*x[3]^4 +x[1]*x[2] -x[1]*x[3] +x[2]*x[3] -2*x[3]^2]

Executing test Curve_h
RingWithID(12, "ZZ/(101)[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = true
Curve_h = [x[0]^3 +100*x[1]^3,  x[0]^2*x[1]^3 +100*x[2]^5,  x[1]^6 +100*x[0]*x[2]^5,  x[0]^2*x[2]^5 +100*x[3]^7,  x[1]^3*x[2]^5 +100*x[0]*x[3]^7,  x[2]^10 +100*x[1]^3*x[3]^7]

Executing test Curve_h
RingWithID(13, "ZZ/(101)[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = true
Curve_h = [x[0]^3 +100*x[1]^3,  x[0]^2*x[1]^3 +100*x[2]^5,  x[1]^6 +100*x[0]*x[2]^5,  x[0]^2*x[2]^5 +100*x[3]^7,  x[1]^3*x[2]^5 +100*x[0]*x[3]^7,  x[2]^10 +100*x[1]^3*x[3]^7]

Executing test Curve_h
RingWithID(14, "QQ[#[5],#[6],#[7],#[8]]")   GradingDim = 1   IsHomog(InputPolys) = true
Curve_h = [#[5]^3 -#[6]^3,  #[5]^2*#[6]^3 -#[7]^5,  #[6]^6 -#[5]*#[7]^5,  #[5]^2*#[7]^5 -#[8]^7,  #[6]^3*#[7]^5 -#[5]*#[8]^7,  #[7]^10 -#[6]^3*#[8]^7]

Executing test Curve_h
RingWithID(15, "QQ[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = true
Curve_h = [x[0]^3 -x[1]^3,  x[0]^2*x[1]^3 -x[2]^5,  x[1]^6 -x[0]*x[2]^5,  x[0]^2*x[2]^5 -x[3]^7,  x[1]^3*x[2]^5 -x[0]*x[3]^7,  x[2]^10 -x[1]^3*x[3]^7]

Executing test Curve
RingWithID(12, "ZZ/(101)[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = false
Curve = [x[0]^3 +100*x[1],  x[0]^2*x[1] +100*x[2],  x[1]^2 +100*x[0]*x[2],  x[0]^2*x[2] +100*x[3],  x[2]^2 +100*x[1]*x[3],  x[1]*x[2] +100*x[0]*x[3]]

Executing test Curve
RingWithID(13, "ZZ/(101)[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = false
Curve = [x[0]^3 +100*x[1],  x[0]^2*x[1] +100*x[2],  x[1]^2 +100*x[0]*x[2],  x[0]^2*x[2] +100*x[3],  x[2]^2 +100*x[1]*x[3],  x[1]*x[2] +100*x[0]*x[3]]

Executing test Curve
RingWithID(14, "QQ[#[5],#[6],#[7],#[8]]")   GradingDim = 1   IsHomog(InputPolys) = false
Curve = [#[5]^3 -#[6],  #[5]^2*#[6] -#[7],  #[6]^2 -#[5]*#[7],  #[5]^2*#[7] -#[8],  #[7]^2 -#[6]*#[8],  #[6]*#[7] -#[5]*#[8]]

Executing test Curve
RingWithID(15, "QQ[x[0],x[1],x[2],x[3]]")   GradingDim = 1   IsHomog(InputPolys) = false
Curve = [x[0]^3 -x[1],  x[0]^2*x[1] -x[2],  x[1]^2 -x[0]*x[2],  x[0]^2*x[2] -x[3],  x[2]^2 -x[1]*x[3],  x[1]*x[2] -x[0]*x[3]]

Executing test Curve
RingWithID(20, "QQ[x[0],x[1],x[2],x[3]]")   GradingDim = 0   
Curve = [x[0]^3 -x[1],  x[0]^2*x[1] -x[2],  x[0]*x[2] -x[1]^2,  x[0]*x[1]^2 -x[3],  x[1]^4 -x[2]*x[3],  x[1]*x[3] -x[2]^2,  x[0]*x[3] -x[1]*x[2],  x[1]^3*x[2] -x[3]^2,  x[1]^2*x[2]^3 -x[3]^3,  x[1]*x[2]^5 -x[3]^4,  x[2]^7 -x[3]^5]

